### PR TITLE
Add global util for exiting dirty form

### PR DESCRIPTION
**Pull request by mmarkusik**
I want to merge this change because it adds exit dirty form prompt

**PR intended to be tested with API branch:** 

### Pull Request Checklist

1. [x] This code contains UI changes
2. [ ] All visible strings are translated with proper context including data-formatting
3. [ ] Attributes `[data-test-id]` are added for new elements
4. [ ] Changes are mentioned in the changelog
5. [ ] The changes are tested in different browsers and in light/dark mode

### Test environment config

API_URI=https://master.staging.saleor.cloud/graphql/

### CODE REVIEW

**Comment by patrys** [comment]
Task linked: [SALEOR-2990 Add global util for exiting dirty form](https://app.clickup.com/t/2549495/SALEOR-2990)

File: src/channels/pages/ChannelDetailsPage/ChannelDetailsPage.tsx [hunk 1]
```diff
@@ -79,7 +79,7 @@ export const ChannelDetailsPage: React.FC<ChannelDetailsPageProps> = ({
     );
 
   return (
-    <Form onSubmit={onSubmit} initial={initialData}>
+    <Form confirmLeave onSubmit={onSubmit} initial={initialData}>
```
Comment (mmarkusik) [inline src/channels/pages/ChannelDetailsPage/ChannelDetailsPage.tsx:82]: Leaving only this form with confirmLeave for QA purposes, all the other places will be added in a separate pr

File: src/hooks/useForm.ts [hunk 2]
```diff
@@ -12,7 +14,7 @@ export interface ChangeEvent<TData = any> {
     value: TData;
   };
 }
-export type SubmitPromise = Promise<any[]>;
+export type SubmitPromise<TData = any> = Promise<TData>;
```
Comment (mmarkusik) [inline src/hooks/useForm.ts:17]: Since our submits are not consistent and not always return an array, leaving this for now and will type all the SubmitPromises properly in the next pr

File: src/utils/richText/useRichText.ts [hunk 3]
```diff
@@ -25,6 +25,10 @@ function useRichText(opts: {
   }, [opts.initial]);
 
   const change: RichTextEditorChange = newData => {
+    if (data.current.blocks.length === 0 && newData.blocks.length === 0) {
```
Comment (mmarkusik) [inline src/utils/richText/useRichText.ts:28]: Added this because useRichText was triggering setChanged to true after a successful submit with nothing changed in the description

File: src/utils/handlers/handleFormSubmit.ts [hunk 4]
```diff
@@ -3,16 +3,26 @@ import { SubmitPromise } from "@saleor/hooks/useForm";
 async function handleFormSubmit<T>(
   data: T,
   onSubmit: (data: T) => SubmitPromise,
-  setChanged: (changed: boolean) => void
+  setChanged: (changed: boolean) => void,
+  setEnableExitPrompt?: (value: boolean) => void
 ): Promise<boolean> {
-  const errors = await onSubmit(data);
-  const ok = errors?.length === 0;
+  const result = onSubmit(data);
```
Comment (jwm0) [inline src/utils/handlers/handleFormSubmit.ts:9]: Didn't we want to wrap this in try/catch blocks and use promise rejection as a mechanism for submit errors?

**Review by jwm0 — CHANGES_REQUESTED** [review]

File: src/utils/handlers/handleFormSubmit.ts (hunk 4 above)
Comment (mmarkusik) [reply src/utils/handlers/handleFormSubmit.ts:9]: Resolving due to this issue being addressed in the next pr

**Review by eCzerniak — APPROVED** [review]

File: src/categories/components/CategoryCreatePage/form.tsx [hunk 5]
```diff
@@ -21,33 +25,37 @@ interface CategoryCreateHandlers {
...
+  privateMetadata: [],
+  seoDescription: "",
+  seoTitle: "",
+  slug: ""
+};
+
 function useCategoryCreateForm(
-  onSubmit: (data: CategoryCreateData) => Promise<any[]>
+  onSubmit: (data: CategoryCreateData) => Promise<any[]>,
+  confirmLeave
```
Comment (jwm0) [inline src/categories/components/CategoryCreatePage/form.tsx]: Didn't we discuss to move this to hook explicitly?

**Review by jwm0 — APPROVED** [review]
